src/astarte_device.c

```diff
@@ -1078,6 +1151,239 @@ static void send_emptycache(astarte_device_handle_t device)
+        if (!interface_is_in_introspection(device, interface_name)) {
+            // Check if this is the last iterable item
+            bool has_next = false;
+            storage_err = astarte_storage_iterator_peek(&storage_iterator, &has_next);
+            if (storage_err != ASTARTE_OK) {
+                ESP_LOGE(TAG, "Error peaking next property.");
+                astarte_storage_close(storage_handle);
+                goto end;
+            }
+            // Delete the property
+            ESP_LOGD(TAG, "Deleting old property '%s%s' from storage", interface_name, path);
+            storage_err = astarte_storage_delete_property(storage_handle, interface_name, path);
+            if ((storage_err != ASTARTE_OK) && (storage_err != ASTARTE_ERR_NOT_FOUND)) {
+                ESP_LOGE(TAG, "Error deleting the property.");
+                astarte_storage_close(storage_handle);
+                goto end;
+            }
+            // If the deleted iterm was the last one, break the loop here
+            if (!has_next) {
+                break;
+            }
+            // Prevent the iterator from advancing
+            advance_iterator = false;
+        }
+        // If property is device owned send it to Astarte
+        else if (interface_is_device_owned(device, interface_name)) {
+            // Assuming that since the property is present in storage the size of the value
+            // has already been checked and does not exceed the max value for an integer.
+            publish_data(device, interface_name, path, value, (int) value_len, 2);
+
+            // Get the combined interface_name and path for the property
+            size_t property_full_path_len = interface_name_len + path_len - 1;
+            char *property_full_path = calloc(property_full_path_len, sizeof(char));
+            strncat(strncpy(property_full_path, interface_name, interface_name_len), path,
+                path_len - 1);
+
+            // Store the property full path in a temporary set
+            astarte_err_t list_err = astarte_linked_list_append(&list_handle, property_full_path);
+            if (list_err != ASTARTE_OK) {
+                ESP_LOGE(TAG, "Error adding a property name to the set %s.",
+                    astarte_err_to_name(list_err));
+                astarte_storage_close(storage_handle);
+                goto end;
+            }
+        }
+        // Free memory of property data
+        free(interface_name);
+        interface_name = NULL;
+        free(path);
+        path = NULL;
+        free(value);
+        value = NULL;
+        // Advance the iterator if required
+        if (advance_iterator) {
+            storage_err = astarte_storage_iterator_advance(&storage_iterator);
+            if (storage_err != ASTARTE_OK) {
+                if (storage_err != ASTARTE_ERR_NOT_FOUND) {
+                    ESP_LOGE(TAG, "Error advancing properties iterator.");
+                }
+                break;
```

Comment:
I think you can combine these lines into a single line, and remove the break line, because the `storage_error` is already checked in the while loop, so you can  simplify this piece of code.
``` c
if (storage_err != ASTARTE_OK && storage_err != ASTARTE_ERR_NOT_FOUND) {
     ESP_LOGE(TAG, "Error advancing properties iterator.");
}
```

Reply:
True